server/src/main/java/org/elasticsearch/action/admin/indices/refresh/TransportShardRefreshAction.java

```diff
@@ -86,43 +84,11 @@ protected void shardOperationOnPrimary(
         IndexShard primary,
         ActionListener<PrimaryResult<BasicReplicationRequest, ReplicationResponse>> listener
     ) {
-        try (var listeners = new RefCountingListener(listener.map(v -> new PrimaryResult<>(shardRequest, new ReplicationResponse())))) {
-            var refreshResult = primary.refresh(SOURCE_API);
+        ActionListener.completeWith(listener, () -> {
+            primaryRefreshResult = primary.refresh(SOURCE_API);
```

Comment:
Transport actions are instantiated as singletons so storing a result specific to a shard in a class attribute can cause bugs. That would be nice if we can use the `PrimaryResult` to hold the primary shard result for the time of the execution

Reply:
Ah good catch! I just realize this. I did not want originally to be touching the replica request (to avoid any bwc issues), but I see now it is necessary. However, I can do something similar we did with the IndexShardRoutingTable parameter in the unpromotable request -- make the primary refresh result a package-private parameter to add in the replica request, and not send/serialize it over the wire. I did that in the latest commit, feel free to review that.